### PR TITLE
Enable splash screen properties for VB WinForms projects.

### PR DESCRIPTION
**Pull request by melytc**
Addresses #8286, part of #8766, [AB#1565607](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1565607)

This PR:
- makes the splash screen properties visible
- creates a `IDynamicEnumProvider` for the splash screen property, where it gets the valid entry points from the `IEntryPointFinderService`.
- saves the splash screen property as the name of the entry point, without any namespaces, as that's the way the myapp file needs it to generate the code for the Designer (in contrast to the C# way we do it for the `StartupObject` property.)
- tracks the `StartupObject` in the `ConfigurationGeneral` xaml rule to later be used in the `IDynamicEnumValuesGenerator` to remove the selected value (if any) from the splash screen enum value list.

**Pending:**
I'll add some unit tests to ensure edge cases are considered.
- Test: change in the startup object value refreshes the splash screen enum value list.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8945)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties.VisualBasic;
+
+[ExportDynamicEnumValuesProvider("SplashScreenEnumProvider")]
+[AppliesTo(ProjectCapability.VisualBasic)]
+internal class SplashScreenEnumProvider : IDynamicEnumValuesProvider
+{
+    private readonly Workspace _workspace;
+    private readonly UnconfiguredProject _unconfiguredProject;
+
+    public SplashScreenEnumProvider([Import(typeof(VisualStudioWorkspace))] Workspace workspace, UnconfiguredProject unconfiguredProject)
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs:18]: For this MEF part to work, you'll need `[ImportingConstructor]` here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        bool includeEmptyValue = options?.Any(pair =>
+            pair.Name == "IncludeEmptyValue"
+            && bool.TryParse(pair.Value, out bool optionValue)
+            && optionValue) ?? false;
+
+        return Task.FromResult<IDynamicEnumValuesGenerator>(new StartupObjectsEnumGenerator(_workspace, _unconfiguredProject, includeEmptyValue, true));
+    }
+}
+
+internal class SplashScreenEnumGenerator : IDynamicEnumValuesGenerator
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs:41]: Looks like you could make this a nested class of `SplashScreenEnumProvider`.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs (hunk 1 above)
Comment (melytc) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs:18]: good catch!

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            if (entryPoints is not null)
+            {
+                enumValues.AddRange(entryPoints.Select(ep =>
+                {
+                    string displayName = ep.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat.WithGlobalNamespaceStyle(SymbolDisplayGlobalNamespaceStyle.Omitted));
+                    return new PageEnumValue(new EnumValue { Name = ep.Name, DisplayName = displayName });
+                }));
+            }
+
+            // Test: change in the startup object value refreshes the splash screen enum value list.
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs:96]: Did you mean to include this comment?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/ApplicationPropertyPage/ApplicationFrameworkValueProvider.cs [hunk 4]
```diff
@@ -254,7 +254,7 @@ private async Task<bool> IsWPFApplicationAsync(IProjectProperties defaultPropert
             return null;
         }
 
-        private async Task<string> GetPropertyValueAsync(string propertyName)
+        private async Task<string> GetPropertyValueAsync(string propertyName, IProjectProperties defaultProperties)
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/ApplicationPropertyPage/ApplicationFrameworkValueProvider.cs:257]: Is anything using the new parameter here?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            }
+
+            IEntryPointFinderService? entryPointFinderService = project.Services.GetService<IEntryPointFinderService>();
+            IEnumerable<INamedTypeSymbol>? entryPoints = entryPointFinderService?.FindEntryPoints(compilation.GlobalNamespace, _searchForEntryPointsInFormsOnly);
+            if (entryPoints is not null)
+            {
+                enumValues.AddRange(entryPoints.Select(ep =>
+                {
+                    string displayName = ep.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat.WithGlobalNamespaceStyle(SymbolDisplayGlobalNamespaceStyle.Omitted));
+                    return new PageEnumValue(new EnumValue { Name = ep.Name, DisplayName = displayName });
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs:92]: It'd be good to document why both name and full name are used here, including what's written in the `.myapp` file.

Also what happens when the user types in a value that's not in the list. My guess is it'd end up in the `.myapp` file with the root namespace prefixed.

Another question is what happens if the `.myapp` file contains a fully qualified name? Does it find the splash screen? If so, would we fail to select it in the UI here, or would we create two entries with the same `DisplayName` but different `Name` values?

File: src/Microsoft.VisualStudio.Editors/MyApplication/MyApplicationCodeGenerator.vb
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.Editors/MyApplication/MyApplicationCodeGenerator.vb:1]: This is the legacy UI yeah? Do we need to update this any more?

Comment (melytc) [reply src/Microsoft.VisualStudio.Editors/MyApplication/MyApplicationCodeGenerator.vb:1]: Only for the `MinimumSplashScreenDisplayTime` property; the `SplashScreen` and `HighDpiMode` properties that we introduced recently already have their code behind 🥳

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/ApplicationPropertyPage/ApplicationFrameworkValueProvider.cs (hunk 4 above)
Comment (melytc) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/ApplicationPropertyPage/ApplicationFrameworkValueProvider.cs:257]: Woop, not anymore, let me take it out

**Review by drewnoakes — APPROVED** [review]

File: src/Microsoft.VisualStudio.Editors/MyApplication/MyApplicationCodeGenerator.vb
Comment (melytc) [reply src/Microsoft.VisualStudio.Editors/MyApplication/MyApplicationCodeGenerator.vb:1]: This is the class that takes the content of the myapp file and generates the necessary code for the VB compiler. Here's an example:
image
image

It is also used by the legacy UI, since we also surface the Application Framework properties over there.

Comment (melytc) [reply src/Microsoft.VisualStudio.Editors/MyApplication/MyApplicationCodeGenerator.vb:1]: The splash screen property already exists in here, but not the MinimumSplashScreenDisplayTime property, so that's the main change here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs (hunk 3 above)
Comment (melytc) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs:96]: Let me remove it and move this reminder in the PR description :)

**Review by MiYanni — APPROVED** [review]

**Review by haileymck — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs (hunk 5 above)
Comment (melytc) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/VisualBasic/SplashScreenEnumProvider.cs:92]: A user can enter a value in the myapp file directly and we will try to show it in the UI as it is, using the `TryCreateEnumValueAsync` method below. So, we will not be adding the root namespace prefix, just have the value be the name and display name for that enum value.

As for the question around having the fully qualified name, the code won't be generated then and we would create two different entries in the UI, the one we picked up and this fully qualified one (we would generate it via the `TryCreateEnumValueAsync`).
